From b081d5ecf4de116c410d833b418f70c34e7d0703 Mon Sep 17 00:00:00 2001
From: Matthias Schoepfer <m.schoepfer@rethinkrobotics.com>
Date: Wed, 8 Mar 2023 13:18:28 +0100
Subject: [PATCH] CMakeLists.txt: fixing relocatable package issue

With the current CMakeLists, the installed package will have an absolute
path, which breaks if the package gets relocated later on.

This patch tries to set up a proper header only ros2 package.

Signed-off-by: Matthias Schoepfer <m.schoepfer@rethinkrobotics.com>
---
 CMakeLists.txt | 27 +++++++++++++++++++++++----
 1 file changed, 23 insertions(+), 4 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 63c1563..1dfde0b 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -13,9 +13,28 @@ endif()
 find_package(ament_cmake REQUIRED)
 find_package(Eigen3 REQUIRED)
 
-install(DIRECTORY include/${PROJECT_NAME}/
-        DESTINATION include/${PROJECT_NAME}
-        FILES_MATCHING PATTERN "*.h")
+add_library(${PROJECT_NAME} INTERFACE
+  include/${PROJECT_NAME}/eigen_stl_containers.h
+  include/${PROJECT_NAME}/eigen_stl_map_container.h
+  include/${PROJECT_NAME}/eigen_stl_vector_container.h
+)
+
+ament_target_dependencies(${PROJECT_NAME} Eigen3)
+ament_export_targets(${PROJECT_NAME}Targets HAS_LIBRARY_TARGET)
+ament_export_dependencies(Eigen3)
+
+install(
+  DIRECTORY include/
+  DESTINATION include
+)
+
+install(
+  TARGETS ${PROJECT_NAME}
+  EXPORT ${PROJECT_NAME}Targets
+  LIBRARY DESTINATION lib
+  ARCHIVE DESTINATION lib
+  RUNTIME DESTINATION bin
+  INCLUDES DESTINATION include
+)
 
-ament_export_include_directories(include ${EIGEN3_INCLUDE_DIRS})
 ament_package()
-- 
2.39.2

